internal/app/machined/pkg/controllers/k8s/apiserver_loadbalancer.go

```diff
@@ -0,0 +1,233 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/channel"
+	"github.com/siderolabs/gen/slices"
+	"github.com/siderolabs/go-loadbalancer/loadbalancer"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+	"github.com/siderolabs/talos/pkg/machinery/resources/k8s"
+)
+
+// APILoadBalancerController creates load balancer based on APIServerEndpointsType resource.
+type APILoadBalancerController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *APILoadBalancerController) Name() string {
+	return "k8s.APILoadBalancerController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *APILoadBalancerController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: k8s.NamespaceName,
+			Type:      k8s.APIServerEndpointsType,
+			ID:        pointer.To(k8s.APIServerEndpointsID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        pointer.To(config.V1Alpha1ID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *APILoadBalancerController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: k8s.LoadBalancerStatusesType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *APILoadBalancerController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	logger = logger.Named("api-endpoints-balancer")
+
+	port, err := ctrl.loadBalancerData(ctx, r, logger)
+	if err != nil {
+		return err
+	} else if ctx.Err() != nil {
+		return nil //nolint:nilerr
+	}
+
+	if port == 0 {
```

Comment:
this is not a way to really have reactive controller - the controller should actively listen for config changes, and bring the lb up/down if the config changes.

Reply:
Yeah - reworked. Done!